### PR TITLE
Reveal Rx pipeline for multi-model environment

### PR DESCRIPTION
**Pull request by j-bjorne**
Still some work to do, but a generic class is up.
Need some more implementation, but after that should be test ready.

### CODE REVIEW

File: viewer/src/data/repository/RxModelSectorsRepository.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  constructor(modelService: ModelSectorService, sectorDataParser: RxModelSectorParser<T>) {
+    this.modelService = modelService;
+    this.sectorDataParser = sectorDataParser;
+  }
+
+  getSectors(): OperatorFunction<ModelSector[], any> {
+    return pipe(
+      mergeAll(20),
+      publish(modelSectorArrayObservable => {
+        const [cachedSectorsObservable, networkSectorsObservable] = partition(
```
Comment (dragly) [inline viewer/src/data/repository/RxModelSectorsRepository.ts:32]: Nice :)

File: viewer/src/data/network/client/ModelSectorService.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+export interface ModelSectorService {
+  fetchModelSector(modelId: number, file: string): Promise<Uint8Array>;
```
Comment (dragly) [inline viewer/src/data/network/client/ModelSectorService.ts:6]: Is this implemented yet?

File: viewer/src/data/repository/RxModelSectorsRepository.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  private readonly sectorCache: { [modelSector: string]: Observable<T> } = {};
+
+  constructor(modelService: ModelSectorService, sectorDataParser: RxModelSectorParser<T>) {
+    this.modelService = modelService;
+    this.sectorDataParser = sectorDataParser;
+  }
+
+  getSectors(): OperatorFunction<ModelSector[], any> {
+    return pipe(
+      mergeAll(20),
```
Comment (dragly) [inline viewer/src/data/repository/RxModelSectorsRepository.ts:30]: Is there a rate limit in place anywhere? If not, what is the effect of mergeAll?

File: viewer/src/data/model/three/RevealObject3D.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+/*!
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/model/three/RevealObject3D.ts:1]: It is there as a placeholder xD

File: viewer/src/data/model/three/RevealObject3D.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+ * Copyright 2020 Cognite AS
+ */
+
+import { Object3D } from 'three';
+
+export class RevealObject3D extends Object3D {}
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/model/three/RevealObject3D.ts:7]: Indeed, RevealObject3D is not a good name. It was added as a placeholder for now. I also don't want to tie it to CAD only.

File: viewer/src/data/network/client/ModelSectorService.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+/*!
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/network/client/ModelSectorService.ts:1]: In the newest version this is now called a ModelSectorProvider, and it is in the folder data/http/provider, the client which implements all the providers is in data/http/client. Data models are the models of data indeed. We can move the network folder out of data, but it is part of the data access layer in my opinion.

File: viewer/src/data/parser/ModelSectorData.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+// Could be separated, I would like the format to be generic and not just cad specific
+
+export enum ModelSectorFormat {
+  i3d,
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/parser/ModelSectorData.ts:8]: Agreed!

File: viewer/src/data/parser/ModelSectorData.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+// Could be separated, I would like the format to be generic and not just cad specific
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/parser/ModelSectorData.ts:5]: Sure. What about Data3D?

File: viewer/src/data/parser/ThreeModelDataParser.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+import { RxModelSectorParser } from './RxModelDataParser';
+import { RevealObject3D } from '../model/three/RevealObject3D';
+import { OperatorFunction, pipe, merge } from 'rxjs';
+import { ModelSectorData, ModelSectorFormat } from './ModelSectorData';
+import { Sector, SectorQuads } from '../../models/cad/types';
+import { publish, filter, map } from 'rxjs/operators';
+
+interface WebAssemblyParser {
```
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/parser/ThreeModelDataParser.ts:12]: Yes. I first called it a RustParser, then changed it to WA parser. The name is currently more to show that we have multiple parsers, we have the one in core, and then one afterwards to change it to the view type.

File: viewer/src/data/network/client/ModelSectorService.ts (hunk 2 above)
Comment (j-bjorne) [reply viewer/src/data/network/client/ModelSectorService.ts:6]: I did add a RevealClient.ts which uses this right?

File: viewer/src/data/repository/RxModelSectorsRepository.ts (hunk 3 above)
Comment (j-bjorne) [reply viewer/src/data/repository/RxModelSectorsRepository.ts:30]: mergeAll() is a limit, but limit is weird in rxjs.

If I say a limit 2, then lower down limit 2, the first one processes two, and those two process two of their own, so 4 concurrent operations are in flight.

File: viewer/src/data/model/three/RevealObject3D.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/data/model/three/RevealObject3D.ts:1]: Ah, no worries then :)

File: viewer/src/data/parser/ModelSectorData.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/data/parser/ModelSectorData.ts:5]: Hm... not sure. Let us keep this open and maybe hear if @larsmoa has some input.

File: viewer/src/data/parser/ThreeModelDataParser.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/data/parser/ThreeModelDataParser.ts:12]: Ah, yes. Should the other one be called Parser when it is not working with file data?

File: viewer/src/data/network/client/ModelSectorService.ts (hunk 2 above)
Comment (dragly) [reply viewer/src/data/network/client/ModelSectorService.ts:6]: Yes, I saw the usage, but not the implementation. Is that added yet?

Comment (dragly) [reply viewer/src/data/network/client/ModelSectorService.ts:6]: Perhaps you forgot to add a file?

File: viewer/src/data/repository/RxModelSectorsRepository.ts (hunk 3 above)
Comment (dragly) [reply viewer/src/data/repository/RxModelSectorsRepository.ts:30]: Ah, that's right. Thanks :)

File: viewer/src/data/parser/ThreeModelDataParser.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (j-bjorne) [reply viewer/src/data/parser/ThreeModelDataParser.ts:12]: Transformer?

File: viewer/src/data/network/client/ModelSectorService.ts (hunk 2 above)
Comment (j-bjorne) [reply viewer/src/data/network/client/ModelSectorService.ts:6]: No, I forgot, I added it last night but didn't push it yet xD I'll add it once I have solved my current brain puzzle

**Pull request closed without merging**